### DISCUSSION TITLE
Inconsistencies with "lib" in external Library names.

### DISCUSSION THREAD

**Discussion by tburch-LSS** [OpenModelica/OpenModelica#12885] [Q&A] [answered]
I've been contributing to the integration of GSL with Modelica and have noticed a seeming inconsistency when naming external libraries within annotation statements. For example, when I use Library="libgsl" I receive a notification that I need to rename the library to "liblibgsl.so" or "liblibgsl.a" (I resolve this with symbolic links). When I then use annotation(Library="gsl"), however, the library is simply not found. I have checked that the LibraryDirectory is set correctly (it is the default one).

Oddly enough, the case is exactly the opposite with the interface library: "gsl_integration_MI" works, but "libgsl_integration_MI" does not... ???

(I am using OMEdit v1.23.1 on Ubuntu 22.04.)

**Comment by casella** [comment]
@adrpo, @sjoelund can you please comment on that?

From what I understand, the leading "lib" is a Linux thing, but I may be wrong.

**Reply by perost** [reply, marked as answer]
I don't know why the library isn't found in this case, but `annotation(Library="gsl")` is the correct form since different platforms have different naming conventions which the compiler takes care of (e.g. `libgsl.so` on Linux and `gsl.dll` on Windows) .

**Reply by sjoelund** [reply]
I'm not sure if we look for symlinks or only regular files. Don't have time to try it though

**Reply by tburch-LSS** [reply]
i was also under the impression that the leading "lib" is a linux thing. hence my surprise when it didn't work for "gsl". it does work for the interface i put together: libgsl_integration_MI.so is found with "gsl_integration_MI".
hmm... i'll keep trying to straighten it out.
given your obvious workload with migration to 1.24.0, i thank you for the responses!

**Reply by tburch-LSS** [reply]
got things working a bit better now...
e.g., with the model shown below annotation(Library="gsl") works fine with the impure function accessing GSL's error function, erf. the function used to access GSL's integration routine QAG contains annotation(Library="gsl_integration_MI") [note: NO mention of "gsl"]. if i had to guess, i would suspect that the inclusion of the latter "gsl" sends omc looking for my system's version of gsl (e.g., on /lib/x86_64-linux-gnu/) and fails to find libgsl_integration_MI.so there. not sure though...

**Reply by tburch-LSS** [reply]
working well enough for now. i'll mark this issue as closed...

**Comment by tburch-LSS** [comment]
the plot thickens...

if i simulate models which only try to access GSL's special functions, then the annotation(Library="gsl") works fine. however, when i try to incorporate the integration routines -- via the libgsl_integration_MI.so interface: annotation(Library=("gsl_integration_MI","gsl")) -- then the GSL integration routines are not found.

however, even stranger, if i change the annotation in the call to the external GSL function, specfunc.erf, from "gsl" to "libgsl", then the integration routines are found and the model shown below works.

i am stumped...

![ExampleIntegral](https://github.com/user-attachments/assets/27d4a3ff-ae4d-4023-b2d3-23c8cc039b29)

Instantiated model (Library annotations not shown):

impure function GNU_ScientificLibrary.Functions.integration.qag "QAG integration routine"
  input Integer n_par;
  input Real[n_par] par;
  input Real a;
  input Real b;
  input Real rel_err;
  input Integer limit;
  input Integer key;
  output Real y;
  output Real err;

  external "C" qag_mi(y, err, par, a, b, rel_err, limit, key);
end GNU_ScientificLibrary.Functions.integration.qag;

impure function GNU_ScientificLibrary.Functions.specfunc.erf "Error function"
  input Real x;
  output Real y;

  external "C" y = gsl_sf_erf(x);
end GNU_ScientificLibrary.Functions.specfunc.erf;

class GNU_ScientificLibrary.Examples.integration.ExampleIntegral "Example integration using GSL"
  final parameter Integer n_par = 1;
  parameter Real a = 0.0;
  parameter Real rel_err = 1e-6;
  parameter Integer limit = 1000;
  parameter Integer key = 6;
  parameter Real b_ramp.height = 0.99 "Height of ramps";
  parameter Real b_ramp.duration(quantity = "Time", unit = "s", min = 0.0, start = 2.0) = 0.99 "Duration of ramp (= 0.0 gives a Step)";
  Real b_ramp.y "Connector of Real output signal";
  parameter Real b_ramp.offset = 0.01 "Offset of output signal y";
  parameter Real b_ramp.startTime(quantity = "Time", unit = "s") = 0.01 "Output y = offset for time < startTime";
  parameter Real I_qag.k(unit = "1", start = 1.0) = 1.1283791670955126 "Gain value multiplied with input signal";
  Real I_qag.u "Input signal connector";
  Real I_qag.y "Output signal connector";
  Real iiratio.u1 "Connector of Real input signal 1";
  Real iiratio.u2 "Connector of Real input signal 2";
  Real iiratio.y "Connector of Real output signal";
  parameter Real a_const.k(start = 1.0) = a "Constant output value";
  Real a_const.y "Connector of Real output signal";
  parameter Real p_const[1].k(start = 1.0) = 1.0 "Constant output value";
  Real p_const[1].y "Connector of Real output signal";
  final parameter Integer qAG_quad.n_par = 1 "Number of parameters in integrand";
  parameter Integer qAG_quad.limit = 1000;
  parameter Integer qAG_quad.key = 6;
  parameter Real qAG_quad.rel_err = 1e-7;
  Real qAG_quad.err;
  Real qAG_quad.a;
  Real qAG_quad.b;
  Real qAG_quad.par[1];
  Real qAG_quad.y;
  Real erf.y;
  Real erf.x;
equation
  I_qag.y = iiratio.u2;
  b_ramp.y = erf.x;
  b_ramp.y = qAG_quad.b;
  p_const[1].y = qAG_quad.par[1];
  a_const.y = qAG_quad.a;
  qAG_quad.y = I_qag.u;
  erf.y = iiratio.u1;
  b_ramp.y = b_ramp.offset + (if time < b_ramp.startTime then 0.0 else if time < b_ramp.startTime + b_ramp.duration then (time - b_ramp.startTime) * b_ramp.height / b_ramp.duration else b_ramp.height);
  I_qag.y = I_qag.k * I_qag.u;
  iiratio.y = iiratio.u1 / iiratio.u2;
  a_const.y = a_const.k;
  p_const[1].y = p_const[1].k;
  (qAG_quad.y, qAG_quad.err) = GNU_ScientificLibrary.Functions.integration.qag(1, qAG_quad.par, qAG_quad.a, qAG_quad.b, qAG_quad.rel_err, qAG_quad.limit, qAG_quad.key);
  erf.y = GNU_ScientificLibrary.Functions.specfunc.erf(erf.x);
end GNU_ScientificLibrary.Examples.integration.ExampleIntegral;
